File: src/main/java/com/wafflestudio/draft/security/GeneralAuthenticationFilter.java
package com.wafflestudio.draft.security;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.wafflestudio.draft.dto.request.AuthenticationRequest;
import com.wafflestudio.draft.security.oauth2.OAuth2Token;
import com.wafflestudio.draft.security.oauth2.client.exception.SucceedOAuthUserNotFoundException;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.AuthenticationException;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.security.web.authentication.UsernamePasswordAuthenticationFilter;
import org.springframework.security.web.util.matcher.AntPathRequestMatcher;

import javax.servlet.FilterChain;
import javax.servlet.ServletException;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.BufferedReader;
import java.io.IOException;

public class GeneralAuthenticationFilter extends UsernamePasswordAuthenticationFilter {

    private JwtTokenProvider jwtTokenProvider;

    public GeneralAuthenticationFilter(AuthenticationManager authenticationManager, JwtTokenProvider jwtTokenProvider) {
        super();
        this.setAuthenticationManager(authenticationManager);
        this.setRequiresAuthenticationRequestMatcher(new AntPathRequestMatcher("/api/v1/user/signin/", "POST"));
        this.jwtTokenProvider = jwtTokenProvider;
    }

    @Override
    protected void successfulAuthentication(HttpServletRequest request, HttpServletResponse response, FilterChain chain, Authentication authResult) {
        response.addHeader("Authentication", jwtTokenProvider.generateToken(authResult));
        response.setStatus(HttpServletResponse.SC_NO_CONTENT);
    }

    @Override
    protected void unsuccessfulAuthentication(HttpServletRequest request, HttpServletResponse response, AuthenticationException failed) throws IOException, ServletException {
//        super.unsuccessfulAuthentication(request, response, failed);
        if (failed instanceof SucceedOAuthUserNotFoundException) {
            response.setStatus(HttpServletResponse.SC_NOT_FOUND);
        } else {
            response.setStatus(HttpServletResponse.SC_UNAUTHORIZED);
        }
    }

    @Override
    public Authentication attemptAuthentication(HttpServletRequest request, HttpServletResponse response) throws AuthenticationException {
        // Parse auth request
        AuthenticationRequest parsedRequest;
        try {
            parsedRequest = parseRequest(request);
        } catch (IOException e) {
            throw new RuntimeException("Bad request");
        }

        String grantType = parsedRequest.getGrantType();

        Authentication authRequest;

        switch (grantType) {
            case "OAUTH":
                authRequest = new OAuth2Token(null, parsedRequest);
                break;
            case "PASSWORD":
                authRequest = new UsernamePasswordAuthenticationToken(parsedRequest.getEmail(), parsedRequest.getPassword());
                break;
            default:
                throw new UsernameNotFoundException(String.format("Grant Type '%s' is not supported", grantType));
        }

        return this.getAuthenticationManager().authenticate(authRequest);
    }

    private AuthenticationRequest parseRequest(HttpServletRequest request) throws IOException {
        BufferedReader reader;

        reader = request.getReader();
        ObjectMapper mapper = new ObjectMapper();
        return mapper.readValue(reader, AuthenticationRequest.class);

    }
}

File: src/main/kotlin/com/wafflestudio/draft/security/GeneralAuthenticationFilter.kt
package com.wafflestudio.draft.security

import com.fasterxml.jackson.databind.ObjectMapper
import com.wafflestudio.draft.dto.request.AuthenticationRequest
import com.wafflestudio.draft.security.oauth2.OAuth2Token
import com.wafflestudio.draft.security.oauth2.client.exception.SucceedOAuthUserNotFoundException
import org.springframework.security.authentication.AuthenticationManager
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.Authentication
import org.springframework.security.core.AuthenticationException
import org.springframework.security.core.userdetails.UsernameNotFoundException
import org.springframework.security.web.authentication.UsernamePasswordAuthenticationFilter
import org.springframework.security.web.util.matcher.AntPathRequestMatcher
import java.io.BufferedReader
import java.io.IOException
import javax.servlet.FilterChain
import javax.servlet.ServletException
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse

class GeneralAuthenticationFilter(authenticationManager: AuthenticationManager?, jwtTokenProvider: JwtTokenProvider) : UsernamePasswordAuthenticationFilter() {
    private val jwtTokenProvider: JwtTokenProvider
    override fun successfulAuthentication(request: HttpServletRequest, response: HttpServletResponse, chain: FilterChain, authResult: Authentication) {
        response.addHeader("Authentication", jwtTokenProvider.generateToken(authResult))
        response.status = HttpServletResponse.SC_NO_CONTENT
    }

    @Throws(IOException::class, ServletException::class)
    override fun unsuccessfulAuthentication(request: HttpServletRequest, response: HttpServletResponse, failed: AuthenticationException) {
//        super.unsuccessfulAuthentication(request, response, failed);
        if (failed is SucceedOAuthUserNotFoundException) {
            response.status = HttpServletResponse.SC_NOT_FOUND
        } else {
            response.status = HttpServletResponse.SC_UNAUTHORIZED
        }
    }

    @Throws(AuthenticationException::class)
    override fun attemptAuthentication(request: HttpServletRequest, response: HttpServletResponse): Authentication {
        // Parse auth request
        val parsedRequest: AuthenticationRequest = try {
            parseRequest(request)
        } catch (e: IOException) {
            throw RuntimeException("Bad request")
        }
        val grantType = parsedRequest.grantType
        val authRequest: Authentication
        authRequest = when (grantType) {
            "OAUTH" -> OAuth2Token(null, parsedRequest)
            "PASSWORD" -> UsernamePasswordAuthenticationToken(parsedRequest.email, parsedRequest.password)
            else -> throw UsernameNotFoundException(String.format("Grant Type '%s' is not supported", grantType))
        }
        return authenticationManager.authenticate(authRequest)
    }

    @Throws(IOException::class)
    private fun parseRequest(request: HttpServletRequest): AuthenticationRequest {
        val reader: BufferedReader
        reader = request.reader
        val mapper = ObjectMapper()
        return mapper.readValue(reader, AuthenticationRequest::class.java)
    }

    init {
        this.authenticationManager = authenticationManager
        setRequiresAuthenticationRequestMatcher(AntPathRequestMatcher("/api/v1/user/signin/", "POST"))
        this.jwtTokenProvider = jwtTokenProvider
    }
}
